gocov/convert/convert.go

```diff
@@ -57,9 +53,43 @@ func ConvertProfiles(filenames ...string) ([]byte, error) {
 		if err != nil {
 			return nil, err
 		}
-		for _, p := range profiles {
-			if err := converter.convertProfile(packages, p); err != nil {
-				return nil, err
+
+		mapUniqPackageNames := make(map[string]interface{})
+		uniqPackageNames := make([]string, 0, len(profiles))
+		for _, profile := range profiles {
+			packageName := filepath.Dir(profile.FileName)
+
+			if _, ok := mapUniqPackageNames[packageName]; ok {
+				continue
+			}
+
+			mapUniqPackageNames[packageName] = nil
+			uniqPackageNames = append(uniqPackageNames, packageName)
+		}
+
+		packages, err := goPackages.Load(&goPackages.Config{
+			Mode: goPackages.NeedName | goPackages.NeedCompiledGoFiles,
+		}, uniqPackageNames...)
+		if err != nil {
+			return nil, fmt.Errorf("load packages: %v", err)
+		}
+
+		for _, profile := range profiles {
+			for _, pkg := range packages {
+				for _, compiledGoFile := range pkg.CompiledGoFiles {
+					// pkg.PkgPath -> 	     github.com/app/internal/service
+					// profile.FileName ->   github.com/app/internal/service/hello.go
+					// strings.TrimPrefix(strings.TrimPrefix(profile.FileName, pkg.PkgPath), "/") -> hello.go
+					// and because in go package we can't have files with similar names we can find absolute path
+					// compiledGoFile -> /usr/admin/repos/app-local/internal/service/hello.go
+					// _, file = filepath.Split(compiledGoFile) -> hello.go
+					_, compiledGoFileName := filepath.Split(compiledGoFile)
+					if strings.TrimPrefix(strings.TrimPrefix(profile.FileName, pkg.PkgPath), "/") == compiledGoFileName {
+						if err := converter.convertPackage(profile, compiledGoFile, pkg.PkgPath); err != nil {
+							return nil, err
+						}
+					}
+				}
```

Comment:
Instead of iterating through each package, could we create a map from pkg.PkgPath -> pkg and do a lookup on that?

i.e.

```go
pkgmap := make(map[string]*goPackages.Package, len(packages))
for _, pkg := range packages {
	pkgmap[pkg.PkgPath] = pkg
}

for _, profile := range profiles {
	pkgpath, filename := path.Split(profile.FileName)
	pkg := pkgmap[pkgpath]
	for _, abspath := range pkg.CompiledGoFiles {
		if filepath.Base(abspath) == filename {
			if err := converter.convertPackage(profile, abspath, pkg.PkgPath); err != nil {
				return nil, err
			}
		}
	}
}
```

Should be a bit more efficient.